Support Azure OpenAI API

sunshine0523 opened the pull request:
Let openai-java support access Azure OpenAI API

axiangcoding commented:
related to https://github.com/TheoKanning/openai-java/issues/249, we are looking forward to this pr being merged

sunshine0523 commented:
> related to #249, we are looking forward to this pr being merged

Yes, I saw this issue, but it seems that Azure OpenAI is not just about changing the baseUrl. Am I misunderstood?

simonlmartin commented:
Also looking forward to this PR being merged... would it be possible to add a constructor which  enables the abiilty to pass in the AzureOpenAiApi class like the OpenAIService does so it is possible to add custom interceptors to the  chain?

iuadastra commented:
@TheoKanning  can this be merged? It's been pending for too long

wilmarDupont commented:
Looking forward to it. It would be incredibly helpful to me.

zhaosen241 commented:
Also looking forward to it, please merge it.

bowenliang123 commented on service/src/main/java/com/theokanning/openai/service/OpenAiService.java:
```diff
@@ -4,10 +4,7 @@
-import com.theokanning.openai.OpenAiApi;
-import com.theokanning.openai.OpenAiError;
-import com.theokanning.openai.OpenAiHttpException;
+import com.theokanning.openai.*;
```
Unnecessary changes to imports of `OpenAiService`.  Unrelated to the purposed feature. Changes should be reverted.

bowenliang123 commented on service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java:
```diff
@@ -0,0 +1,224 @@
+     *
+     * @param apiCall The api call
+     */
+    public static Flowable<SSE> stream(retrofit2.Call<ResponseBody> apiCall) {
```
```suggestion
    public static Flowable<SSE> stream(Call<ResponseBody> apiCall) {
```
Simplify it to `Call`, as `retrofit2.Call` is already imported.

bowenliang123 commented on service/src/main/java/com/theokanning/openai/service/AzureOpenAiService.java:
```diff
@@ -0,0 +1,224 @@
+     * @param apiCall  The api call
+     * @param emitDone If true the last message ([DONE]) is emitted
+     */
+    public static Flowable<SSE> stream(retrofit2.Call<ResponseBody> apiCall, boolean emitDone) {
```
ditto

sunshine0523 replied:
I have also done other related work, but this pull request has not been merged, so I did not submit the new work

bowenliang123 replied:
What you mentioned sounds not a blocker here to simplify the reference here. Each PR should be standalone and good to reflect changes to the purpose.

bowenliang123 commented on client/src/main/java/com/theokanning/openai/AzureOpenAiApi.java:
```diff
@@ -0,0 +1,75 @@
+import retrofit2.Call;
+import retrofit2.http.*;
+
+public interface AzureOpenAiApi{
```
```suggestion
public interface AzureOpenAiApi {
```

Format the code please.

bowenliang123 commented:
I am worried about the too much duplicated code to `OpenAiService` in this feature.
Not a good practice in software design patterns. Also bad in maintenance.
In this way, we will have many different similar versions with duplication for different vendors of Open AI services.

sunshine0523 commented:
> I am worried about the too much duplicated code to `OpenAiService` in this feature.
> Not a good practice in software design patterns. Also bad for future maintenance.
> In this way, we will have many different similar versions with duplication for different vendors of Open AI services.

I agree with you, I changed the library for my own demo. But now I've changed the implementation. The implementation is similar to OpenAI's Python library. I think that way is good.

qixiaobo commented:
Maybe next release?

Hakenadu commented:
I'd also love to see this merged, as otherwise we would have to implement our own client for this purpose.
Are there any plans for this?

jnunderwood commented:
Is there a status update for this PR? We could really use it. Thank you!

bowenliang123 commented:
Any response or update to the review comments in the unnecessary code changes, the worries about duplicate code as existing `OpenAiService`? I don't think it's in a good enough shape.

sunshine0523 commented:
I think this code was written very poorly, so I closed this pull request

bowenliang123 commented:
At least the minor code suggestions should be good to be updated.
As `open-java` lib is shipped with Apache Kyuubi to provide ChatGPT as the chatting engine, we do concern about the quality and maintenance of the upstream dependencies.

sunshine0523 commented:
> At least the minor code suggestions should be good to be updated. As `open-java` lib is shipped with Apache Kyuubi to provide ChatGPT as the chatting engine, we do concern about the quality and maintenance of the upstream dependencies.

This project has not been updated for some time. Currently, there are other libraries that support Azure OpenAI. Perhaps you can consider replacing the active maintenance project.

df007df commented:
> > At least the minor code suggestions should be good to be updated. As `open-java` lib is shipped with Apache Kyuubi to provide ChatGPT as the chatting engine, we do concern about the quality and maintenance of the upstream dependencies.
> 
> This project has not been updated for some time. Currently, there are other libraries that support Azure OpenAI. Perhaps you can consider replacing the active maintenance project.

你倒是说用哪个可以
